File: python/448.find_all_numbers_disappeared_in_an_array.py
from typing import List


class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:

        result = []

        for num in nums:
            idx = abs(num) - 1
            nums[idx] = abs(nums[idx]) * -1

        for index, num in enumerate(nums):
            if num > 0:
                result.append(index + 1)

        return result

File: cpp/448.find_all_numbers_disappeared_in_an_array.cpp
#include <vector>

class Solution {
 public:
    std::vector<int> findDisappearedNumbers(std::vector<int>& nums) {
        std::vector<int> result;

        for (auto num : nums) {
            // since each (number -1) in the array corresponds to an index
            // we keep flipping numbers to negative for each idx = num - 1
            int i = abs(num) - 1;
            nums[i] = -1 * abs(nums[i]);
        }

        int i = 0;
        for (auto num : nums) {
            if (num > 0) {
                result.push_back(i + 1);
            }
            ++i;
        }
        return result;
    }
};
